src/coreclr/jit/codegenarm64.cpp

```diff
@@ -1853,8 +1853,17 @@ void CodeGen::genCodeForBinary(GenTreeOp* treeNode)
 
     // The arithmetic node must be sitting in a register (since it's not contained)
     assert(targetReg != REG_NA);
+    emitAttr attr = emitActualTypeSize(treeNode);
 
-    regNumber r = emit->emitInsTernary(ins, emitActualTypeSize(treeNode), treeNode, op1, op2);
+    // UMULL/SMULL is twice as fast for 32*32->64bit MUL
+    if (oper == GT_MUL && EA_SIZE(attr) == EA_8BYTE && !varTypeIsFloating(treeNode->TypeGet()) &&
+        EA_SIZE(emitActualTypeSize(op1)) == EA_4BYTE && EA_SIZE(emitActualTypeSize(op2)) == EA_4BYTE)
+    {
+        ins  = (treeNode->gtFlags & GTF_UNSIGNED) ? INS_umull : INS_smull;
+        attr = EA_4BYTE;
+    }
+
+    regNumber r = emit->emitInsTernary(ins, attr, treeNode, op1, op2);
```

Comment:
```suggestion
    // UMULL/SMULL is twice as fast for 32*32->64bit MUL.
    if ((oper == GT_MUL) && (targetType == TYP_LONG) && genActualTypeIsInt(op1) && genActualTypeIsInt(op2))
    {
        ins  = treeNode->IsUnsigned() ? INS_umull : INS_smull;
        attr = EA_4BYTE;
    }

    regNumber r = emit->emitInsTernary(ins, attr, treeNode, op1, op2);
```
Just a suggestion on how to write the same code using a bit terser APIs.

Reply:
I like the code suggested by @SingleAccretion.
@pentp can you please consider the change?